### PR TITLE
[FIX] microsoft_calendar: missing attendees confirmation

### PR DESCRIPTION
**Pull request by geflx**
Before this commit, when a user was invited to an event from an account outside Outlook (e.g. Google account), accepting the invitation in Outlook Calendar didn't send the 'accept' status to Odoo Calendar, and vice-versa. This problem was also happening with recurrent events accepted in Outlook, the events weren't being accepted in Odoo Calendar correctly.

After this commit, both ways are fixed: accepting an event in Outlook Calendar or Odoo Calendar notifies the other calendar correctly. For recurrent events, accepting them in Outlook Calendar also updates the events status to 'accept' in Odoo Calendar (reminder: due to a limitation in Outlook Calendar synchronization, when the synchronization is active recurrent events acceptance can only be done through Outlook).

Task-id: 3290922

### CODE REVIEW

File: addons/microsoft_calendar/models/calendar.py [hunk 1]
```diff
@@ -217,6 +222,18 @@ def _get_organizer_user_change_info(self, values):
                     partner_ids.append(command[2].get('partner_id'))
         return sender_user, partner_ids
 
+    def _update_attendee_status(self, attendee_ids):
+        # Merge current status from attendees_ids with new attendees values for avoiding their info loss in write().
+        # Create a dict getting the state of each attendee received from attendee_ids variable and then update their state.
+        state_by_attendee = {}
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/calendar.py:228]: You're not grouping by attendee, but by partner
```suggestion
        state_by_partner = {}
```

File: addons/microsoft_calendar/models/calendar.py [hunk 2]
```diff
@@ -217,6 +222,18 @@ def _get_organizer_user_change_info(self, values):
...
+        for cmd in attendee_ids:
+            if len(cmd) == 3 and isinstance(cmd[2], dict) and all(key in cmd[2] for key in ['partner_id', 'state']):
+                state_by_attendee[cmd[2]['partner_id']] = cmd[2]['state']
+        for attendee in self.attendee_ids:
+            state_update = state_by_attendee.get(attendee.partner_id.id)
+            if state_update:
+                attendee.write({'state': state_update})
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/calendar.py:235]:
```suggestion
                attendee.state = state_update
```

File: addons/microsoft_calendar/models/calendar.py [hunk 3]
```diff
@@ -171,7 +171,7 @@ def write(self, values):
                 self._forbid_recurrence_update()
 
         # When changing the organizer, check its sync status and verify if the user is listed as attendee.
-        if values.get('user_id') and self.user_id.id != values['user_id']:
+        if self and values.get('user_id') and self.user_id.id != values['user_id']:
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/calendar.py:174]: Is there any write where you don't have self set?

File: addons/calendar/static/src/views/attendee_calendar/attendee_calendar_model.js [hunk 4]
```diff
@@ -94,7 +94,7 @@ export class AttendeeCalendarModel extends CalendarModel {
                     }
                     // Records will share the same rawRecord.
                     const record = { ...event };
-                    const attendeeInfo = data.attendees.find(a => (
+                    const attendeeInfo = data.attendees.findLast(a => (
```
Comment (leonardopavanrocha) [inline addons/calendar/static/src/views/attendee_calendar/attendee_calendar_model.js:97]: Why is this change required?

File: addons/microsoft_calendar/models/calendar.py [hunk 5]
```diff
@@ -272,6 +289,13 @@ def _microsoft_to_odoo_values(self, microsoft_event, default_reminders=(), defau
             'show_as': 'free' if microsoft_event.showAs == 'free' else 'busy',
             'recurrency': microsoft_event.is_recurrent()
         })
+        # Edge case: do not add 'start' and 'stop' variables when set by Microsoft while they are the same as in Odoo.
+        # Otherwise, the return will be considered as 'update_time' in 'write' and set attendees state as 'needsAction'.
+        update_edge_case = self.env.context.get('update_edge_case', False)
+        if update_edge_case and self.start == start and self.stop == stop:
+            values.pop('start')
+            values.pop('stop')
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/calendar.py:297]: I had a similar case in google_calendar and solved it by not setting start/stop in the `_odoo_values` in google calendar (`_microsoft_to_odoo_values` in microsoft) https://github.com/odoo/odoo/pull/131966/files#diff-7d087533ebb0faba0b7d853b15b2ca5f9e3113bdd9b64c0eea14d14229bf1002R123-R126.  Maybe it's better to fix it like that rather than having to pass stuff in the context.

File: addons/microsoft_calendar/models/microsoft_sync.py [hunk 6]
```diff
@@ -378,7 +378,7 @@ def _sync_microsoft2odoo(self, microsoft_events: MicrosoftEvent):
                 ms_event_updated_time = parse(mevent.lastModifiedDateTime)
 
                 if ms_event_updated_time >= odoo_event_updated_time:
-                    vals = dict(odoo_event._microsoft_to_odoo_values(mevent), need_sync_m=False)
+                    vals = dict(odoo_event.with_context(update_edge_case=True)._microsoft_to_odoo_values(mevent), need_sync_m=False)
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/microsoft_sync.py:381]: update_edge_case seems a bit weird. Because the edge case is the combination of this context value as truthy with the update of start/stop. This context value represents only that the ms_event is more up to date than the odoo event.

File: addons/microsoft_calendar/models/microsoft_sync.py [hunk 7]
```diff
@@ -453,13 +453,16 @@ def _microsoft_attendee_answer(self, answer, params, timeout=TIMEOUT):
...
-                    answer, params, token=token, timeout=timeout
-                )
-                self.write({
-                    'need_sync_m': False,
-                })
+                status, event = microsoft_service._get_single_event(self.ms_universal_event_id, token=token)
+                if status and event and event.get('value'):
+                    # Update the event's iCalUId as it changes for each attendee.
+                    # Send respective attendee answer with updated iCalUId.
+                    self.ms_organizer_event_id = event.get('value')[0].get('id')
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/microsoft_sync.py:460]: Are you guaranteed to always have a value, it being an array with an object in the first position having an id as a key? If so, you don't need to use `.get`. Even now, you're using `.get`, but if it doesn't return anything, it will break when trying to read the 0th position.

File: addons/microsoft_calendar/models/microsoft_sync.py [hunk 8]
```diff
@@ -453,13 +453,16 @@ def _microsoft_attendee_answer(self, answer, params, timeout=TIMEOUT):
...
+                    res = microsoft_service.answer(
+                        event.get('value')[0].get('id'),
+                        answer, params, token=token, timeout=timeout
+                    )
+                    self.write({'need_sync_m': not res})
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/models/microsoft_sync.py:465]:
```suggestion
                    self.need_sync_m = not res
```

File: addons/microsoft_calendar/tests/test_answer_events.py [hunk 9]
```diff
@@ -27,51 +28,46 @@ def setUp(self):
                     microsoft_id=combine_ids("123", "456"),
                 )
             )
+        for user in [self.organizer_user, self.attendee_user]:
+            user.microsoft_calendar_token_validity = datetime.now() + timedelta(hours=1)
```
Comment (leonardopavanrocha) [inline addons/microsoft_calendar/tests/test_answer_events.py:32]:
```suggestion
        (self.orgarnizer_user | self.attendee_user).microsoft_calendar_token_validity =  datetime.now() + timedelta(hours=1)
```

**Review by leonardopavanrocha — CHANGES_REQUESTED** [review]
Thanks for your work! I have some comments/questions. Could you check this when you have some time?

File: addons/microsoft_calendar/models/microsoft_sync.py (hunk 6 above)
Comment (leonardopavanrocha) [reply addons/microsoft_calendar/models/microsoft_sync.py:381]: Yes, or 'ms_event_is_most_recent'

File: addons/microsoft_calendar/models/calendar.py (hunk 3 above)
Comment (leonardopavanrocha) [reply addons/microsoft_calendar/models/calendar.py:174]: Can't we do an early return if self is empty? I guess calling this without a self won't cause any side effects

File: addons/calendar/static/src/views/attendee_calendar/attendee_calendar_model.js (hunk 4 above)
Comment (leonardopavanrocha) [reply addons/calendar/static/src/views/attendee_calendar/attendee_calendar_model.js:97]: We should check this. Creating a new attendee for every update seems like an error

**Review by arj-odoo — CHANGES_REQUESTED** [review]
I have some questions/blocking point on this one. I think we are near a solution but we must be careful :-).

**Review by arj-odoo — CHANGES_REQUESTED** [review]
I have still a few remarks. I just want to be sure we don't introduce any issue.

**Comment by arj-odoo** [comment]
@robodoo r+
